HW3/P5/label_regions.cl

```diff
@@ -79,21 +82,65 @@ propagate_labels(__global __read_write int *labels,
     // the pixel for this thread
     old_label = buffer[buf_y * buf_w + buf_x];
 
-    // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+    // Calculate Pix 
+    // Initialize Grandfather Label
+    int prev_label=0;
+    
+    // Start at the First Pixel (0, 0) location in local buffer
+    if (idx_1D == 0){   
+        
+        // Store grandparent for comparsion
+        if (old_label < w*h)
+            prev_label = labels[old_label];
+
+        // Loop through the local work group minus halo 
+        // ind = [y * buf_w + x]
+        for (int ind=1; ind < ((buf_w-halo) * (buf_h-halo)); ind++) {
```

Comment:
The indexing here is wrong.
To avoid the halo values you should start from buf_w + halo, and make sure that every buf_w-2 elements you increase the index by 2 (look at the 1D indices of an inner rectangle inside a rectangle).